### PR TITLE
Add "any" rule for branching validations

### LINKED ISSUE #53: Support branching
I want to be able to validate that something is either undefined or a number, for instance. Is there already a way to do that? If not, it might be worth adding.

### PR DESCRIPTION
**Pull request by imbrn**
This is a preliminary implementation and should be validated yet.

We have to make some decisions like what we should expect when no validation is given. Should we get a valid or invalid result?

In the current implementation, I decided to make the arguments spread to an array `(... validations)` instead of getting an array directly. What do you think about?

And this PR closes #53.

### CODE REVIEW

**Comment by sebastianbarfurth** [comment]
Your choice of passing if either passes but failing if both fail seems like the best choice. However, consequently this means that no given validation should in fact fail since none passed. It’s an edge case regardless, I believe whatever decision will be ok.

This rule is confusingly named in regards to to the some modifier. I don’t know a better name.

I like the implementation all in all.

**Comment by imbrn** [comment]
I started implementing this by considering the result invalid when no validations were given at all. But I ended up with this current implementation. But I think I should go back to the first choice.

About the naming, I also think it's confusing with the `some` modifier. But I can't think of another name, too.

I'll wait a little while until merging it, anyway. So if someone has a suggestion, it will be appreciated. 😁

**Comment by imbrn** [comment]
What do you think about `passesAny` for the name? It's showing in a better way what it's going to do. And it prevents us from confusing with `some` and `every` modifiers, I think.

```js
// some validated value must pass any of the specified validations
v8n().some.passesAny(
	v8n().number(),
	v8n().null()
).test(["foo", "bar", 3]); // true

// every validated value must pass any of the specified validations
v8n().every.passesAny(
	v8n().number(),
	v8n().null()
).test([1, 2, null]); // true
```

**Comment by sebastianbarfurth** [comment]
For actual fluent language, why not use `passesAnyOf`? I think that way it would sound really nice and also stand out from the modifiers. In the long run I feel like an `or` modifier is the best solution, but that can wait until this is implemented in a basic way.

**Comment by imbrn** [comment]
`passesAnyOf` sounds better indeed.

File: src/v8n.js [hunk 1]
```diff
@@ -152,7 +152,8 @@ const availableRules = {
 
   // branching
 
-  any: (...validations) => value => validations.some(it => it.test(value))
+  passesAnyOf: (...validations) => value =>
+    validations.some(it => it.test(value))
```
Comment (sebastianbarfurth) [inline src/v8n.js]: `it` is not a descriptive variable name, why not `validation`?

**Comment by sebastianbarfurth** [comment]
Needs to be added to CHANGELOG.

**Comment by iangilman** [comment]
Fantastic! Thank you for implementing this :-)

One question: how does this behave when you use `check` instead of `test`? I don't see that covered in the test suite. I presume the exception just says it failed the "passesAnyOf"?

**Comment by sebastianbarfurth** [comment]
It uses `test` internally, so if any validation fails inside `passesAnyOf` it will in fact throw a `ValidationException` that has the `rule` attribute `passesAnyOf` since that's what is being validated with `check`.

Maybe this could be improved to reflect the failed branch.

You could open an issue for this, I don't believe this is strictly related to the rule itself, it's more of a stylistic choice.

**Comment by imbrn** [comment]
It makes no sense to reflect a failed branch because all branches must fail for the `passesAnyOf` rule to fail.

So if your `passesAnyOf` rule has failed, you can get the failed rules from the `ex.rule.args` property.

**Comment by iangilman** [comment]
Sorry for the late reply here, but thank you both for the explanation!

**Pull request merged**